feat(developer): add user interface for `kmc copy` to TIKE

mcdurdin opened the pull request:
Adds a new set of 'New Project' dialogs to allow authors to easily copy keyboards from local filesystem, Keyman Cloud (including a keyboard search web page, implemented in keymanapp/keyman.com#536), and directly from a GitHub URL. Adds related documentation.

Clone from local filesystem:
![image](https://github.com/user-attachments/assets/40a3dc63-eb62-43bb-809a-fd35b1594231)

Clone from Keyman Cloud:
![image](https://github.com/user-attachments/assets/8f4eb097-3246-4ec4-b394-097129a01a12)

Clone from GitHub:
![image](https://github.com/user-attachments/assets/637dfadf-c63f-45f1-98ac-372c1956edc2)

Fixes: #13073
Depends-on: keymanapp/keyman.com#536

# User Testing

Please note that user test of TEST_CLONE_CLOUD is blocked until keymanapp/keyman.com#536 is merged.

For each of these tests, follow the steps to clone the project, using a new keyboard/lexical model id of your choice, and then make sure that the newly created project builds successfully.

* TEST_CLONE_LOCAL: Try and clone a keyboard project from the local filesystem using Project|New Project|Clone Local Project.
* TEST_CLONE_MODEL_LOCAL: Try and clone a lexical model project from the local filesystem using Project|New Project|Clone Local Project.
* TEST_CLONE_CLOUD: Try and clone a Keyman Cloud keyboard project using Project|New Project|Clone Project from Keyman Cloud. For example, try cloning sil_euro_latin or another well-known keyboard project.
* TEST_CLONE_GITHUB: Try and clone a keyboard project from a GitHub URL, using Project|New Project|Clone Project from GitHub. For example, use 'https://github.com/keymanapp/keyboards/tree/master/release/g/ghana/ghana.kpj'
* TEST_CLONE_MODEL_GITHUB: Try and clone a lexical model project from a GitHub URL, using Project|New Project|Clone Project from GitHub. For example, use 'https://github.com/keymanapp/lexical-models/blob/master/release/av/av.eo.esperanto/av.eo.esperanto.kpj'

dinakaranr commented:
# Test Results
I tested this issue with the attached "Keyman-18.0.178-alpha-test-13076" build(30/01/2025) on Windows 10. I'm sharing my observation here.
* **TEST_CLONE_LOCAL (OPEN):**
1. Open the Keyman Developer application.
2. Open the "Clone local project" dialog by clicking the project -> New Project --> Clone local project.
3. "Source project filename": Select the keyboard *.kpj file from the Local GitHub clone directory.
4. Enter the project name in the "New project ID" text box.
5. Click the "Ok" button on the dialog.
It works well. Thank you.

* **TEST_CLONE_MODEL_LOCAL (Failed):**
1. Open the Keyman_Developer application.
2. Open the "Clone local project" dialog by clicking the project -> New Project --> Clone local project.
3. "Source project filename": Select the keyboard *.kpj file from the local GitHub clone directory.(lexical-models)
4. Enter the project name in the "New project ID" text box.
5. Click the "Ok" button on the dialog.
Actual Results: Here, an Error line appeared in the message box. (I have entered the project name in lowercase. but still, I am getting an error)
Please take a look below screenshot
<img width="300" src="https://github.com/user-attachments/assets/b6acbbfb-279a-48f7-aff7-8a28131f65ea">

* **TEST_CLONE_CLOUD (Failed):**
1. Open the Keyman_Developer application.
2. Open the "Clone project from Keyman Cloud" dialog by clicking the project -> New Project --> Clone project from Keyman Cloud.
Actual Results: Here, the Search box does not appear instead of "Not Found" message appears.
Please take a look below screenshot
<img width="300" src="https://github.com/user-attachments/assets/2740d269-9f8b-4022-b6cf-ef891d3e33df">

* **TEST_CLONE_GITHUB (OPEN):**
1. Open the Keyman_Developer application.
2. Open the "Clone project from GitHub" dialog by clicking the project -> New Project --> Clone project from GitHub.
3. "GitHub URL": Paste the keyboard's GitHub URL
4. Enter the project name in the "New project ID" text box.
5. Click the "Ok" button on the dialog.
It works well. Thank you.

* **TEST_CLONE_MODEL_GITHUB (Failed):**
1. Open the Keyman_Developer application.
2. Open the "Clone project from GitHub" dialog by clicking the project -> New Project --> Clone project from GitHub.
3. "GitHub URL": Paste the keyboard's GitHub URL(Lexical model)
4. Enter the project name in the "New project ID" text box.
5. Click the "Ok" button on the dialog.
Actual Results: Here, an Error line appeared in the message box. (I have entered the project name in lowercase. but still, I am getting an error)
Please take a look below screenshot
<img width="300" src="https://github.com/user-attachments/assets/dcf1eac5-6c86-47ad-abeb-ce9d659520c5">

mcdurdin commented on developer/docs/help/context/clone-keyman-cloud-project-parameters.md:
```diff
@@ -0,0 +1,37 @@
+
+Projects can also be cloned from the command line with [`kmc copy`](kmc).
+
+Only keyboard projects can be cloned from Keyman Cloud at this time via this
```
Good questions. The legacy keyboard would generate an error in the kmc-copy log, so the user would get feedback that way. An open source multi-keyboard package should copy fine -- the .kpj and .kps would get the new id, but the keyboards within would retain their original names. Would be worth a test for each of these, and so I'll give that a go.

mcdurdin commented on developer/src/tike/project/Keyman.Developer.UI.Project.UfrmNewProject.dfm:
```diff
@@ -42,18 +42,26 @@ inherited frmNewProject: TfrmNewProject
     ReadOnly = True
     SmallImages = ilSmall
     TabOrder = 0
+    ViewStyle = vsList
```
I am now using the `vsList` `ViewStyle` because that way the full caption for each option is visible. Previously, the captions were truncated. See the screenshot at https://github.com/keymanapp/keyman/blob/46fc4d52cf500d8a55fae028729218acc0483fd1/developer/docs/help/images/ui/frmNewProject.png (I think the GitHub preview was shrinking it)

mcdurdin commented on developer/src/tike/project/Keyman.Developer.UI.Project.UfrmCloneGitHubProjectParameters.pas:
```diff
@@ -0,0 +1,340 @@
+   * Matches only a GitHub permanent raw URI with a commit hash, without any other
+   * components; note hash is called branch to match other URI formats
+   *}
+  GITHUB_STABLE_SOURCE = '^https:\/\/github\.com\/(?<owner>[a-zA-Z0-9-]+)\/(?<repo>[\w\.-]+)\/raw\/(?<branch>[a-f0-9]{40})\/(?<path>.+)$';
```
It isn't (currently) used but included for completeness to match the typescript source

mcdurdin commented on developer/src/tike/project/Keyman.Developer.UI.Project.UfrmCloneLocalProjectParameters.pas:
```diff
@@ -0,0 +1,262 @@
+
+  if Result then
+  begin
+    if not DirectoryExists(editPath.Text) then
```
No, that's tested in `EnableControls`

mcdurdin commented on developer/src/tike/project/Keyman.Developer.UI.Project.UfrmCloneLocalProjectParameters.pas:
```diff
@@ -0,0 +1,262 @@
+
+procedure TfrmCloneLocalProjectParameters.SetKeyboardID(const Value: string);
+begin
+  editKeyboardID.Text := Value;
```
No, because that's not coming from user input

darcywong00 reviewed (approved):
lgtm

rc-swag commented on developer/docs/help/context/new-project.md:
```diff
@@ -21,4 +21,25 @@ Creates a keyboard project that matches the folder structure used in the Keyman
 
-The **Basic** and **Import Windows Keyboard** modes can also be driven by command line with [KMConvert](kmconvert).
\ No newline at end of file
+## Clone local project
+
+Copies a Keyman keyboard or lexical model project on your local filesystem into
+a new folder, giving the project and included files a new name.
+
+## Clone project from Keyman Cloud
+
+Copies a Keyman keyboard or lexical model project from keyman.com into a new
+folder, giving the project and included files a new name.
+
+## Clone project from GitHub
+
+Copies a Keyman keyboard or lexical model project from a GitHub repository into
+a new folder, giving the project and included files a new name.
```
+1

rc-swag commented on developer/src/tike/compile/Keyman.Developer.System.KmcWrapper.pas:
```diff
@@ -24,18 +34,80 @@ implementation
+  else
+    cmdline := cmdline + ['--no-compiler-warnings-as-errors'];
+
+  if FGlobalProject.Options.WarnDeprecatedCode then
```
```suggestion
  if not FGlobalProject.Options.WarnDeprecatedCode then
```

mcdurdin commented:
@dinakaranr thank you for running these tests. I have updated the user interface to clarify that lexical model IDs should match the pattern [`author.bcp47.uniq`](https://help.keyman.com/developer/lexical-models/#toc-how-the-repository-is-organized)

The test TEST_CLONE_CLOUD was dependent on keymanapp/keyman.com#536 being merged, and that has now been merged.

Can you please re-test all given the changes that I have implemented? Thank you!

@keymanapp-test-bot retest

rc-swag reviewed (approved):
lgtm

dinakaranr commented:
# Test Results
I tested this issue with the attached "Keyman-18.0.178-alpha-test-13076" build(31/01/2025) on Windows 10. I'm sharing my observation here.
* **TEST_CLONE_LOCAL (Passed):**
1. Open the Keyman Developer application.
2. Open the "Clone local project" dialog by clicking the project -> New Project --> Clone local project.
3. "Source project filename": Select the keyboard *.kpj file from the Local GitHub clone directory.
4. Enter the project name in the "New project ID" text box.
5. Click the "Ok" button on the dialog.
6. Verified that the project copied and local directory. It opened in a new window.
It works well. Thank you.

* **TEST_CLONE_MODEL_LOCAL (Passed):**
1. Open the Keyman_Developer application.
2. Open the "Clone local project" dialog by clicking the project -> New Project --> Clone local project.
3. "Source project filename": Select the keyboard *.kpj file from the local GitHub clone directory.(lexical-models)
4. Enter the project name in the "New project ID" text box.
5. Click the "Ok" button on the dialog.
6. Verified that the project copied and local directory. It opened in a new window.
7. It works well. Thank you.

* **TEST_CLONE_CLOUD (Passed):**
1. Open the Keyman_Developer application.
2. Open the "Clone project from Keyman Cloud" dialog by clicking the project -> New Project --> Clone project from Keyman Cloud.
3. "Clone project from Keyman Cloud" appeared.
4. Enter the keyboard in the search box.
5. Enter the project name in the "New project ID" text box.(in lowercase)
6. Verified that the project was downloaded and copied in the local directory. It opened in a new window.
7. It works well. Thank you.
Note: I tried six projects. four projects succeeded and two projects failed.(Message: could not be downloaded)
Please refer to the attached text file.
[Scenario3and5.txt](https://github.com/user-attachments/files/18614620/Scenario3and5.txt)

* **TEST_CLONE_GITHUB (Passed):**
1. Open the Keyman_Developer application.
2. Open the "Clone project from GitHub" dialog by clicking the project -> New Project --> Clone project from GitHub.
3. "GitHub URL": Paste the keyboard's GitHub URL
4. Enter the project name in the "New project ID" text box.
5. Click the "Ok" button on the dialog.
6. Verified that the project was downloaded and copied in the local directory. It opened in a new window.
It works well. Thank you.

* **TEST_CLONE_MODEL_GITHUB (Passed):**
1. Open the Keyman_Developer application.
2. Open the "Clone project from GitHub" dialog by clicking the project -> New Project --> Clone project from GitHub.
3. "GitHub URL": Paste the keyboard's GitHub URL(Lexical model)
4. Enter the project name in the "New project ID" text box.
5. Click the "Ok" button on the dialog.
6. Verified that the project was downloaded and copied in the local directory. It opened in a new window.
It works well. Thank you.
Note: I tried six projects. four projects succeeded and two projects failed. (Message: could not be downloaded)
Please refer to the attached text file.

mcdurdin commented:
> 7\. Note: I tried six projects. four projects succeeded and two projects failed.(Message: could not be downloaded)
>     Please refer to the attached text file.

These messages had the following pattern

`info KM0B01B: The Keyman binary file 'keymanapp/keyboards/tree/master//release/rac/rac_marwari/build/rac_marwari.kmx' could not be downloaded: Error downloading https://raw.githubusercontent.com/keymanapp/keyboards/refs/heads/master/release/rac/rac_marwari/build/rac_marwari.kmx HTTP error 404: Not Found. This is not normally a problem`

The important thing to note is 'This is not normally a problem'. Looking up the reference for the error message KM0B01B shows us:

```
$ kmc message b01b

INFO_CannotDownloadBinaryFileFromGitHub
* Code:      KM0B01B
* Module:    kmc-copy.CopierMessages
* Message:   The Keyman binary file '<param>' could not be downloaded: <param> <param>. This is not normally a problem
* Reference: https://kmn.sh/km0b01b

In most repositories, Keyman binary files such as .kmx, .kmp, .js are not included.
This is not normally a problem, as the files can be built from the source. Check
the provided error details for more details.
```

keyman-server commented:
Changes in this pull request will be available for download in [Keyman version 18.0.185-alpha](https://keyman.com/downloads/releases/alpha/18.0.185)
